Add Agentic Coding Tile (#25)
Ada-Developers-Academy/independent-ai-topics: agentic-coding/mcp-rag.md

> # Adding Outside Knowledge: MCP & RAG

Author: Swapping order so the main subjects of the lesson are visible when the longer name gets truncated.

```suggestion
# MCP & RAG: Adding Outside Knowledge
```

| Tool Schema | The structured definition of a tool's name, description, and expected inputs and outputs, provided by an MCP server. | Tool spec | "The tool schema specified that the `create_ticket` tool required a title and a priority level as inputs." |
| Chunk | A smaller piece of a larger document, broken up so it can be individually embedded and retrieved by a RAG system. | Passage, segment | "The onboarding guide was split into chunks by section, so retrieval could return just the relevant section instead of the whole guide." |
| Embedding | A numeric representation of a piece of text that captures its meaning, used to compare how similar two pieces of text are. | Vector representation | "The retrieval system converted our question into an embedding so it could be compared against the embeddings stored for each document chunk." |
| Vector Database | A database designed to store embeddings and quickly find the ones most similar to a given query. | Embedding store | "The support team's RAG system stores their help articles in a vector database so relevant chunks of articles can be retrieved in milliseconds." |

## Reaching External Tools: MCP Servers

> An **MCP server** is a program that exposes a set of tools to an agent using a shared, open protocol. Instead of every AI tool vendor building its own custom way for agents to talk to Slack, GitHub, or a database, MCP defines one standard so a server built once can be used by any compatible agent.

Reviewer: Although the spelled out name for MCP is given in intro, and in the vocabulary, since this is the first place in the main lesson text that mentions MCP, can we expand the acronym once here?

```suggestion
An **MCP server** is a program that exposes a set of tools to an agent using a shared, open protocol.

**MCP** (Model Context Protocol) is a protocol standardizing how AI applications (built on models) connect to external data sources and tools, so that relevant information can be pulled into a model's context on demand. It also defines how an agent can call a tool and receive its results back.

So instead of every AI tool vendor building its own custom way for agents to talk to Slack, GitHub, or a database, MCP defines one standard so a server built once can be used by any compatible agent.
```

Connecting to an MCP server looks a little different depending on where that server runs, but the underlying steps are the same in either case.

### Local Servers

A local server runs as a process on our own machine, alongside whatever agent tool we're using. We typically set one up by editing a configuration file that tells our agent host what command to run to start the server. That configuration usually specifies:
- A name for the server, so we can identify it in our tool's interface
- The command used to launch it, along with any arguments it must be supplied to run successfully
- Any environment variables the server needs, like an API key it uses to reach a service on our behalf

> Once that configuration is saved and our agent host is restarted, it starts the server process and communicates with it using standard input and output, often referred to as **stdio**.

Reviewer: explicitly provides a reading for stdio
```suggestion
Once that configuration is saved and our agent host is restarted, it starts the server process and communicates with it using standard input and output, often referred to as **stdio** (usually read as either standard input output or standard I O).
```

  - **`"/Users/username/Desktop"` and `"/Users/username/Downloads"`**: These last two arguments aren't flags for `npx`, they're arguments given to the filesystem server when it launches. These are read to determine which directories the server is allowed to touch. Every argument from this point in the array onward is handled by the server's logic rather than by `npx`.

A couple notes about this server and its definition:

- **The directory paths are also the security boundary.** This server will only be able to read or modify files inside `/Users/username/Desktop` and `/Users/username/Downloads`, because that's what its own tools were configured to allow. Adding or removing paths here is how we widen or narrow what the server can reach, without needing to change any code.
- **Nothing here mentions our agent or model at all.** This configuration is entirely about how to launch and scope a program that happens to speak MCP. The agent doesn't know these tools exist until our host restarts and completes tool discovery against this running server.

### !callout-info

## Try it out!

Let's connect a local MCP server in VS Code and confirm that its tools show up where we expect.

> 1. Inside a project run **MCP: Open User Configuration** from the Command Palette to open the MCP configuration file.

Reviewer: I ran into a little bit of a different flow. I had my own sample MCP server configured, so I'm not sure how much of the difference is related to that.

1. Since I already had a server defined, I had to merge the filesystem entry into what I already had. I'm not sure what a user with no configured servers sees.
2. I was not prompted to confirm trust when saving the file.
3. When I opened the Configure Tools option, I saw an unchecked entry for `filesystem`, not a checked entry for `secure-filesystem-server`.
4. There was an entry to Update Servers in the list. Clicking that updated the `filesystem` entry to `secure-filesystem-server`. I still needed to check it manually.
5. When I tried to back out and do it again to confirm the flow, though I still was not prompted on save, when opening the Configure Tools, it showed the checked `secure-filesystem-server` directly, without me needing to click any update option or check the box myself.

I'd be hesitant to describe this exactly in the text as something they might run into, but maybe adding in some flexibility that the exact prompts they receive might differ and that they may encounter an option to explicitly tell VS Code to update its services as part of the flow?

Author: I added some content to the steps below to try to address your concerns, does that match what you were hoping to convey?

Reviewer: Yep, this helps. I think acknowledging there can be some difference from what folks see goes a long way towards encouraging folks to press on through and figure things out. If the directions are very specific, then as soon as what I see differs from what's written, I have to step out a bit and try to figure out whether what I'm seeing is still compatible with what the directions describe. Even a small amount of wiggle room immediately allows me to focus more on making progress than trying to figure out whether I did something wrong.

2. Add a `filesystem` server entry using the configuration we walked through above by copy & pasting the configuration then updating the directory paths (`"/Users/username/Desktop"`, `"/Users/username/Downloads"`) to point at folders on our own machine.
3. Save the file. VS Code will ask us to confirm that we trust the server before it starts, since local servers can run code on our machine.
4. Open the Chat view and select **Configure Tools** in the chat input. We should see the tools VS Code discovered from our new server listed there.
    ![VS Code chat window showing the "Configure Tools" button](assets/mcp-rag/vscode_chat_configure_tools_button.png)   
    *Fig. "Configure Tools" button in the VS Code Chat UI*

    !["secure-filesystem-server" tools showing in the VS Code "Configure Tools" menu](assets/mcp-rag/vscode_tools_dropdown.png)
    *Fig. `secure-filesystem-server` tools showing in the VS Code "Configure Tools" menu*

> 5. Give the agent a prompt that would use one of those tools, such as asking it to list the files in one of the directories we configured, and confirm we're prompted to approve the tool call before it runs.

Reviewer: When I tried this, it looks like VS Code asserts its own directory "roots" which overrides the directories listed in the configuration, causing access to those locations to fail (only the project workspace becomes available). I tried a few workarounds with no success, unfortunately.

_(Interestingly, in the docs for the filesystem mcp, it sounds like they're deprecating the root usage)_

Is there maybe a more innocuous server we could use as an example that's less likely to be impeded by the security model? I know this would entail updating quite a lot of the prior material (json explanations, screenshots, steps).

Or can we compare notes and see whether you run into this now, and if not, see where my settings might differ?

Reviewer: I also looked through the other sample MCP servers available at [modelcontextprotocol](https://github.com/modelcontextprotocol) andthe rest don't seem to suit the bill of having a local MCP server that can be started without any extra installation (i.e., npx) but that can also accept parameters to adjust its scope/permissions.

I was thinking maybe we could fork the server and modify it to ignore the root notification, but then students would have to manualy download and set it up to run on their systems rather than using npx, which seems like a lot more trouble.

Author: The server seems to work for the paths VS Code allows access to (so just the workspace XD), so when I scope the access in the MCP server definition to my working directory, I can invoke the tools just fine. 

```
{
  "servers": {
    "filesystem": {
      "command": "npx",
      "args": [
        "-y",
        "@modelcontextprotocol/server-filesystem",
        "/Users/kelseysteven/Documents/playground/AI/agentic_research/adagrams-py"
      ]
    }
  }
}
```

<img width="642" height="871" alt="Image" src="https://github.com/user-attachments/assets/943f29ba-6011-4394-a821-6ea98c308c1d" />

<img width="637" height="789" alt="Image" src="https://github.com/user-attachments/assets/4f41e9e4-9ab3-4a8d-b776-d2eea64fc29d" />

Something I'm seeing is I kind of have to tell it to use the server tools explicitly. Understandably so, the agent seems to default to the built in tools rather than prioritizing the MCP server when I want it to do something like list files. It means it's an even smaller example than before, but it can still show off adding a local server and invoking tools from it. 

How does that land for you?

- Using our IDE's UI to add a new connection, which will vary from tool to tool

Because the server isn't running under our own user account, most remote servers require an authentication step, commonly OAuth or an API key, before they'll accept requests on our behalf. Remote servers communicate over HTTP rather than stdio, allowing them to be accessed from anywhere with an internet connection.

#### Remote MCP Configuration

Let's take a look at our MCP server configuration from earlier, this time with a remote server connection added to the top of the `servers` list. 

The new connection points at [GitHub's hosted MCP server](https://github.com/github/github-mcp-server), giving an agent tools for working with repositories, issues, and pull requests:

```json
{
  "servers": {
    "github": {
      "type": "http",
      "url": "https://api.githubcopilot.com/mcp/"
    },
    "filesystem": {
      "command": "npx",
      "args": [
        "-y",
        "@modelcontextprotocol/server-filesystem",
        "/Users/username/Desktop",
        "/Users/username/Downloads"
      ]
    }
  }
}
```

> Here's what is happenning in the newly added code:

Reviewer: ```suggestion
Here's what is happening in the newly added code:
```

How we actually interact with RAG systems is typically through MCP, which is why we've presented these topics together. 
- A RAG system creates the infrastructure to provide relevant knowledge, and we interact with that system either as a locally running service, or remotely through an API.
- Whether local or remote, in order for our agents to use a RAG knowledge base, we would connect to that running service using MCP as mentioned in the section above.

### Practical Scenario

A company's legal team maintains thousands of pages of internal compliance policies. When someone asks an agent whether a new vendor contract needs a specific data-handling clause, a RAG system searches that policy library and returns the handful of paragraphs most relevant to the question. The model then answers using those specific paragraphs, rather than guessing based on general knowledge of what compliance policies tend to say. 
- RAG is useful here because contract terms are specific to the organization; there's no way general training data would reflect all of a company's current policies.

> We'll go deeper into RAG in later cloud-focused materials, for now, the key idea to hold onto is that RAG's job is to provide specific knowledge, relevant to a prompt, at the time it is requested.

Reviewer: I don't know if we can really say "deeper", but we can definitely telegraph that we'll be coming back to it.
```suggestion
We'll revisit RAG in later cloud-focused materials. For now, the key idea to hold onto is that RAG's job is to provide specific knowledge, relevant to a prompt, at the time it is requested.
```

- **Reach for RAG over stuffing full documents into context.** If we find ourselves in a pattern of needing to reference a large document or set of documents frequently, and we end up pasting the entire reference document into a conversation when we only need a single section or page, that is a strong sign that a retrieval setup could serve us better long term.

### !callout-info

## Try it out!

Let's look at how to disable tools from a server, shut down a local MCP server, and remove the server entirely, so we know how to disable tools we aren't actively using.

First, we'll check that the server is running: 
1. Run `MCP: List Servers` from the Command Palette  
2. Look for the `filesystem` server in the dropdown list. It will say "Running" next to the name if it is currently active.
    - If the server says "Stopped", click on it and choose the "Start Server" action.

To disable tools from a server without stopping the local server:
1. Open the Chat view and select "**"Configure Tools**"

> 2. Uncheck the box next to `secure-filesystem-server` to disable all tools avilable on the server. 

Reviewer: ```suggestion
2. Uncheck the box next to `secure-filesystem-server` to disable all tools available on the server. 
```

**MCP** (Model Context Protocol) is a protocol standardizing how AI applications (built on models) connect to external data sources and tools, so that relevant information can be pulled into a model's context on demand. It also defines how an agent can call a tool and receive its results back.

So instead of every AI tool vendor building its own custom way for agents to talk to Slack, GitHub, or a database, MCP defines one standard so a server built once can be used by any compatible agent.

Connecting to an MCP server looks a little different depending on where that server runs, but the underlying steps are the same in either case.

### Local Servers

A local server runs as a process on our own machine, alongside whatever agent tool we're using. We typically set one up by editing a configuration file that tells our agent host what command to run to start the server. That configuration usually specifies:
- A name for the server, so we can identify it in our tool's interface
- The command used to launch it, along with any arguments it must be supplied to run successfully
- Any environment variables the server needs, like an API key it uses to reach a service on our behalf

> Once that configuration is saved and our agent host is restarted, it starts the server process and communicates with it using standard input and output, often referred to as **stdio** (usually read as either standard input output or standard I O).

Author: ```suggestion
Once that configuration is saved and our agent host is restarted, it starts the server process and communicates with it using standard input and output, often referred to as **stdio** (usually read as either "standard input output" or "standard I O").
```
